File: modules/face-detector/android/src/main/java/expo/modules/facedetector/FaceDetectorModule.kt
package expo.modules.facedetector

import android.net.Uri
import com.google.mlkit.vision.common.InputImage
import com.google.mlkit.vision.face.FaceDetection
import com.google.mlkit.vision.face.FaceDetectorOptions
import expo.modules.kotlin.Promise
import expo.modules.kotlin.exception.CodedException
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

/**
 * Android face detection with Google ML Kit. Boxes come back in image
 * pixels; they are normalized here so the JS side gets the same top-left
 * fractions Apple Vision produces on iOS.
 */
class FaceDetectorModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("FaceDetector")

    AsyncFunction("detect") { url: String, promise: Promise ->
      val context = appContext.reactContext
        ?: return@AsyncFunction promise.reject(CodedException("ERR_FACE_DETECTION", "No context", null))
      val image = InputImage.fromFilePath(context, Uri.parse(url))
      val options = FaceDetectorOptions.Builder()
        .setPerformanceMode(FaceDetectorOptions.PERFORMANCE_MODE_FAST)
        .build()
      val width = image.width.toDouble()
      val height = image.height.toDouble()
      FaceDetection.getClient(options).process(image)
        .addOnSuccessListener { faces ->
          promise.resolve(faces.map { face ->
            mapOf(
              "x" to face.boundingBox.left / width,
              "y" to face.boundingBox.top / height,
              "width" to face.boundingBox.width() / width,
              "height" to face.boundingBox.height() / height,
            )
          })
        }
        .addOnFailureListener { error ->
          promise.reject(CodedException("ERR_FACE_DETECTION", error.message, error))
        }
    }
  }
}

File: modules/face-detector/ios/FaceDetectorModule.swift
import CoreML
import ExpoModulesCore
import Vision

/**
 * iOS face detection with Apple's Vision framework — no ML Kit pod, works on
 * the simulator. Vision returns boxes normalized to the image with a
 * BOTTOM-left origin, so only the y axis needs flipping to match the
 * top-left convention the app (and ML Kit on Android) uses.
 */
public class FaceDetectorModule: Module {
  public func definition() -> ModuleDefinition {
    Name("FaceDetector")

    AsyncFunction("detect") { (url: URL, promise: Promise) in
      let request = VNDetectFaceRectanglesRequest { request, error in
        if let error {
          promise.reject("ERR_FACE_DETECTION", "Vision failed: \((error as NSError).description)")
          return
        }
        let observations = request.results as? [VNFaceObservation] ?? []
        promise.resolve(observations.map { observation -> [String: Double] in
          let box = observation.boundingBox
          return [
            "x": box.origin.x,
            "y": 1 - box.origin.y - box.size.height,
            "width": box.size.width,
            "height": box.size.height,
          ]
        })
      }

      #if targetEnvironment(simulator)
        // The simulator has no Neural Engine; without forcing the CPU,
        // Vision's face detector can fail with an opaque inference error.
        if #available(iOS 17.0, *) {
          let cpu = MLComputeDevice.allComputeDevices.first { device in
            if case .cpu = device { return true }
            return false
          }
          if let cpu {
            request.setComputeDevice(cpu, for: .main)
          }
        } else {
          request.usesCPUOnly = true
        }
      #endif

      DispatchQueue.global(qos: .userInitiated).async {
        do {
          try VNImageRequestHandler(url: url).perform([request])
        } catch {
          promise.reject("ERR_FACE_DETECTION", "Vision failed: \((error as NSError).description)")
        }
      }
    }
  }
}
